Execute a `rails` smoke test

seanpdoyle opened the pull request:
A smoke test would verify that the `rails` command is properly
configured for a given environment.

This commit also adds `RAILS_ENV={staging,production}` to the test
matrix to ensure that a Suspender-ed application can be run in any Rails
environment without additional configuration.

seanpdoyle commented:
@tute would this sort of testing guard againts regressions for the annoying behavior we experienced today?

tute commented:
> would this sort of testing guard againts regressions for the annoying behavior we experienced today?

I think so, yes. CI fails with an SMTP variable not being there, but it is in `templates/dotfiles/.env`. We might be missing the `bin/setup` command, or you might already have found a regression.

Thank you!

tute commented on .travis.yml:
```diff
@@ -9,3 +9,5 @@ before_install:
 notifications:
   email: false
+env:
+  - RAILS_ENV=production
```
We need to add both `production` and `test` to run both in Travis.

tute commented:
To make `.env` work in production, we'll need to suffix it: `.env.production`.

tute commented:
Marking as "help wanted", so we solve this issue in a Workshop we are about to host (unless you get to it earlier!).

seanpdoyle commented:
@tute related: https://github.com/thoughtbot/suspenders/pull/742

If a smoke test like this were in place, issues like #742 would be caught in CI.

tute commented:
Yep. I don't have time now to move this forward, but I certainly want it in suspenders!

tute commented:
Maybe we get this working with @robskrob!

seanpdoyle commented:
I've rebased and updated this PR.

The main failure right now involves all the `SMTP_*` variables requried by [`smtp.rb`](https://github.com/thoughtbot/suspenders/blob/ffa20dbeec382d6a78704be15881d5dd1248d9df/templates/smtp.rb), as they're not defined by default in `RAILS_ENV={staging,production}`


Some of my ideas:

* Replace the calls to `#fetch` with `#[]` (this is my least favorite option)
* Conditionally set values on the `SMTP_SETTINGS` constant depending on what's available
* Modify the `app.json` somehow to generate garbage values for the variables, and somehow notify the users that they'll need to be updated.

I'm not sure how to proceed.

@tute what do you think?

tute commented on templates/rails_helper.rb:
```diff
@@ -1,4 +1,4 @@
-ENV["RACK_ENV"] = "test"
+ENV["RAILS_ENV"] ||= "test"
```
As of https://github.com/thoughtbot/suspenders/pull/713 we use `RACK_ENV` only. Same goes for `.travis.yml`.

Dorian replied:
If the tests are run in `development` env then this would prevent loading gems in the `test` gemset.

tute commented:
> The main failure right now involves all the SMTP_* variables requried by smtp.rb, as they're not defined by default in RAILS_ENV={staging,production}

We discussed this before. See https://github.com/thoughtbot/suspenders/pull/532#issuecomment-86967089.

seanpdoyle commented:
> suspenders new is not "plug and play"

Does this mean that this PR is no longer applicable?

If we should still proceed with this PR, what do you think about modifying it to define `SMTP_*` variables in the `{staging,production}` portions of the `.travis.yml` matrix, along with matching entries in the project's `.env` file (not the suspendered template)?

tute commented:
Here's a related discussion: https://github.com/thoughtbot/suspenders/pull/692/files#r49132045

I'll give this PR more thought later in the afternoon, Sean.

tute commented:
> Does this mean that this PR is no longer applicable?

This might be the case. Is it a feature that a push to staging with no configured SMTP fails?

```
test-apr-15 master % git push staging
Counting objects: 144, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (126/126), done.
Writing objects: 100% (144/144), 31.15 KiB | 0 bytes/s, done.
Total 144 (delta 11), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Ruby app detected
remote: -----> Compiling Ruby/Rails
remote: -----> Using Ruby version: ruby-2.3.0
remote: -----> Installing dependencies using bundler 1.11.2
remote:        Running: bundle install --without development:test --path vendor/bundle --binstubs vendor/bundle/bin -j4 --deployment
remote:        Fetching gem metadata from https://rubygems.org/...........
remote:        Fetching version metadata from https://rubygems.org/...
remote:        Fetching dependency metadata from https://rubygems.org/..
remote:        Using json 1.8.3
remote:        Installing i18n 0.7.0
remote:        Installing minitest 5.8.4
remote:        Installing rake 11.1.2
remote:        Installing builder 3.2.2
remote:        Installing thread_safe 0.3.5
remote:        Installing erubis 2.7.0
remote:        Installing mini_portile2 2.0.0
remote:        Installing rack 1.6.4
remote:        Installing mime-types-data 3.2016.0221
remote:        Installing arel 6.0.3
remote:        Installing execjs 2.6.0
remote:        Installing addressable 2.4.0
remote:        Installing babel-source 5.8.35
remote:        Installing thor 0.19.1
remote:        Installing sass 3.4.22
remote:        Installing concurrent-ruby 1.0.1
remote:        Installing flutie 2.0.0
remote:        Installing high_voltage 3.0.0
remote:        Installing honeybadger 2.5.3
remote:        Installing normalize-rails 3.0.3
remote:        Installing newrelic_rpm 3.15.1.316
remote:        Installing pg 0.18.4 with native extensions
remote:        Installing puma 3.4.0 with native extensions
remote:        Installing rack-timeout 0.4.2
remote:        Using bundler 1.11.2
remote:        Installing rails_stdout_logging 0.0.5
remote:        Installing tilt 2.0.2
remote:        Installing tzinfo 1.2.2
remote:        Installing nokogiri 1.6.7.2 with native extensions
remote:        Installing rack-test 0.6.3
remote:        Installing rack-mini-profiler 0.9.9.2
remote:        Installing mime-types 3.0
remote:        Installing autoprefixer-rails 6.3.6
remote:        Installing uglifier 3.0.0
remote:        Installing babel-transpiler 0.7.0
remote:        Installing rack-canonical-host 0.2.1
remote:        Installing bourbon 5.0.0.beta.5
remote:        Installing sprockets 3.6.0
remote:        Installing mail 2.6.4
remote:        Installing activesupport 4.2.6
remote:        Installing neat 1.7.4
remote:        Installing sprockets-es6 0.9.0
remote:        Installing rails-deprecated_sanitizer 1.0.3
remote:        Installing recipient_interceptor 0.1.2
remote:        Installing globalid 0.3.6
remote:        Installing activemodel 4.2.6
remote:        Installing delayed_job 4.1.1
remote:        Installing activejob 4.2.6
remote:        Installing activerecord 4.2.6
remote:        Installing delayed_job_active_record 4.1.0
remote:        Installing loofah 2.0.3
remote:        Installing rails-dom-testing 1.0.7
remote:        Installing rails-html-sanitizer 1.0.3
remote:        Installing actionview 4.2.6
remote:        Installing actionpack 4.2.6
remote:        Installing sprockets-rails 3.0.4
remote:        Installing railties 4.2.6
remote:        Installing actionmailer 4.2.6
remote:        Installing jquery-rails 4.1.1
remote:        Installing simple_form 3.2.1
remote:        Installing rails 4.2.6
remote:        Installing sass-rails 5.0.4
remote:        Installing title 0.0.5
remote:        Bundle complete! 45 Gemfile dependencies, 64 gems now installed.
remote:        Gems in the groups development and test were not installed.
remote:        Bundled gems are installed into ./vendor/bundle.
remote:        Bundle completed (30.82s)
remote:        Cleaning up the bundler cache.
remote: -----> Preparing app for Rails asset pipeline
remote:        Running: rake assets:precompile
remote:        rake aborted!
remote:        KeyError: key not found: "SMTP_ADDRESS"
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/config/smtp.rb:2:in `fetch'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/config/smtp.rb:2:in `<top (required)>'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/config/environments/production.rb:1:in `<top (required)>'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:598:in `block (2 levels) in <class:Engine>'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:597:in `each'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:597:in `block in <class:Engine>'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:30:in `instance_exec'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:30:in `run'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:55:in `block in run_initializers'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:44:in `each'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:44:in `tsort_each_child'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:54:in `run_initializers'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/application.rb:352:in `initialize!'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/config/environment.rb:2:in `<top (required)>'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/application.rb:328:in `require_environment!'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/railties-4.2.6/lib/rails/application.rb:457:in `block in run_tasks_blocks'
remote:        /tmp/build_f0d9085e324cd2fc0f471f2a8a299b6f/vendor/bundle/ruby/2.3.0/gems/sprockets-rails-3.0.4/lib/sprockets/rails/task.rb:62:in `block (2 levels) in define'
remote:        Tasks: TOP => environment
remote:        (See full trace by running task with --trace)
remote:  !
remote:  !     Precompiling assets failed.
remote:  !
remote:
remote:  !     Push rejected, failed to compile Ruby app
remote:
remote: Verifying deploy...
remote:
remote: !       Push rejected to test-apr-15-staging.
remote:
To https://git.heroku.com/test-apr-15-staging.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'https://git.heroku.com/test-apr-15-staging.git'
```

tute commented:
> Conditionally set values on the SMTP_SETTINGS constant depending on what's available

As stated in https://github.com/thoughtbot/suspenders/pull/532#issuecomment-86967089, I'm not against conditional values, so that we can configure SMTP later, when needed. Some people at thoughtbot disagree.

seanpdoyle commented:
@tute the motivation behind this PR was making it easier to debug `{staging,production}` issues locally.

The problem is that a newly suspenders app can't `RACK_ENV=production foreman start` without error.

Would it make sense to add a `.env.production` file with garbage for `SMTP` variables? Would Heroku ignore those values?

tute commented:
> Would it make sense to add a `.env.production` file with garbage for `SMTP` variables? Would Heroku ignore those values?

`dotenv-rails` gem is only included in development and test, so this might be a good solution.

seanpdoyle commented:
@mike-burns I haven't considered this code in years, and don't plan to re-visit it.

I'm going to close it.
